pkg/controller/storagecluster/reconcile.go

```diff
@@ -287,6 +312,76 @@ func (r *ReconcileStorageCluster) reconcileNodeTopologyMap(sc *ocsv1.StorageClus
 	return nil
 }
 
+// ensureNodeRacks iterates through the list of storage nodes and ensures
+// all nodes have a rack topology label.
+func (r *ReconcileStorageCluster) ensureNodeRacks(nodes *corev1.NodeList, minRacks int, nodeRacks, topologyMap *ocsv1.NodeTopologyMap, reqLogger logr.Logger) error {
+
+	for _, node := range nodes.Items {
+		hasRack := false
+
+		for nodeName := range nodeRacks.Labels {
+			if nodeName == node.Name {
+				hasRack = true
+				break
+			}
+		}
+
+		if !hasRack {
+			rack := determinePlacementRack(node.Name, minRacks, nodeRacks)
+			reqLogger.Info("Labeling node with rack label", "Node", node.Name, "Label", defaults.RackTopologyKey, "Value", rack)
+			node.Labels[defaults.RackTopologyKey] = rack
+			nodeRacks.Add(node.Name, rack)
+			if !topologyMap.Contains(defaults.RackTopologyKey, rack) {
+				reqLogger.Info("Adding topology label from node", "Node", node.Name, "Label", defaults.RackTopologyKey, "Value", rack)
+				topologyMap.Add(defaults.RackTopologyKey, rack)
+			}
+
+			reqLogger.Info("Updating node", "Node:", node.Name)
+			err := r.client.Update(context.TODO(), &node)
```

Comment:
Patch() is a safer option for objects like this that we didn't create.

There's some gotchas here worth pointing out that aren't immediately obvious. If the node object has new fields that the version of the k8s api we're vendoring doesn't have yet, then when we perform the Update(), it's possible for information to get lost. This is most common for the Status section of objects like this. 

With a Patch() operation, we're explicitly only altering the exact fields we intend to add/remove/mutate. There isn't the same risk of unintentionally altering the object in other ways with a Patch.

Reply:
While I agree here, I don't see that `r.client` has a `Patch()` method. I only see `Create()`, `Delete()`, and `Update()`. Should I be using something else?

Reply:
i see what you mean. They did that because their `r.client` needs the full object in order to determine the object type. There's some magic behind the scenes they're doing to abstract way some of the complexity with the CRD clients. It just doesn't work with Patch.

Well, you could just define your own k8s client in the reconcile object and use it for the node Patch() operation.

I have code in the ```pkg/deploy-manager/config.go``` file that shows how to do get a k8s client.

```
func NewDeployManager() (*DeployManager, error) {
        codecs := serializer.NewCodecFactory(scheme.Scheme)
        parameterCodec := runtime.NewParameterCodec(scheme.Scheme)

        kubeconfig := os.Getenv("KUBECONFIG")
        if kubeconfig == "" {
                return nil, fmt.Errorf("No KUBECONFIG environment variable set")
        }

        // K8s Core api client
        config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
        if err != nil {
                return nil, err
        }
        config.NegotiatedSerializer = serializer.DirectCodecFactory{CodecFactory: scheme.Codecs}
        config.APIPath = "/apis"
        config.ContentType = runtime.ContentTypeJSON
        k8sClient, err := kubernetes.NewForConfig(config)
        if err != nil {
                return nil, err
        }

```

Then there's this code in the ```pkg/deploy-manager/storagecluster.go``` file that shows how to add a label using a strategic merge patch. 

```
        for _, node := range nodes.Items {
                old, err := json.Marshal(node)
                if err != nil {
                        return err
                }
                new := node.DeepCopy()
                new.Labels["cluster.ocs.openshift.io/openshift-storage"] = ""

                newJSON, err := json.Marshal(new)
                if err != nil {
                        return err
                }

                patch, err := strategicpatch.CreateTwoWayMergePatch(old, newJSON, node)
                if err != nil {
                        return err
                }

                _, err = t.k8sClient.CoreV1().Nodes().Patch(node.Name, types.StrategicMergePatchType, patch)
                if err != nil {
                        return err
                }

                labeledCount++
        }

```

Reply:
i know it seems like a pain, but using a patch here saves us the headache of accidentally being responsible for screwing up the node status somehow in the future.